src/librustc_middle/ty/instance.rs

```diff
@@ -502,6 +483,92 @@ impl<'tcx> Instance<'tcx> {
     }
 }
 
+fn polymorphize<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    substs: SubstsRef<'tcx>,
+) -> SubstsRef<'tcx> {
+    debug!("polymorphize({:?}, {:?})", def_id, substs);
+    let unused = tcx.unused_generic_params(def_id);
+    debug!("polymorphize: unused={:?}", unused);
+
+    if unused.is_empty() {
+        // Exit early if every parameter was used.
+        return substs;
+    }
+
+    // If this is a closure or generator then we need to handle the case where another closure
+    // from the function is captured as an upvar and hasn't been polymorphized. In this case,
+    // the unpolymorphized upvar closure would result in a polymorphized closure producing
+    // multiple mono items (and eventually symbol clashes).
+    let upvars_ty = if tcx.is_closure(def_id) {
```

Comment:
is there a reason you aren't doing this as part of `unused_generic_params`?

Reply:
I don't think that would work; `unused_generic_params` could inspect the upvar tuple (it doesn't currently) and then invoke itself on any closure/generator upvars but if it went on to change that upvar closure (like this function does) then that wouldn't change anything outside of the query - by changing this here, it changes what goes into the `MonoItem`.

Reply:
I had a misconception on what we are actually doing here :+1: